### PR TITLE
enhancement(sources): Emit new `source_lag_time_seconds` histogram

### PR DESCRIPTION
**Pull request by bruceg**
This sets up `SourceSender` to emit a new lag time metric for all sources, tagged with the usual source tags.

I know we had discussed the metric being `lag_time_ms`, but the histograms we generate for internal metric are geared towards fractional types, with the top bucket being 4096, which would limit our range to only 4 seconds.

### CODE REVIEW

**Comment by bruceg** [comment]
Postponing this to see if I can make this better with a metrics registry handle through #13625

File: src/source_sender/mod.rs [hunk 1]
```diff
@@ -74,11 +85,13 @@ impl SourceSender {
             buf_size: CHUNK_SIZE,
             inner: None,
             named_inners: Default::default(),
+            lag_time: Some(register_histogram!(LAG_TIME_NAME)),
```
Comment (tobz) [inline src/source_sender/mod.rs:88]: Only caveat is that the call to `SourceSender::builder` needs to happen once we're inside the span for the source component. As long as that's happening, the registered metric will have the right metric labels from the span fields, etc.

File: src/source_sender/mod.rs [hunk 2]
```diff
@@ -74,11 +85,13 @@ impl SourceSender {
...
         }
     }
 
     pub fn new_with_buffer(n: usize) -> (Self, LimitedReceiver<EventArray>) {
-        let (inner, rx) = Inner::new_with_buffer(n, DEFAULT_OUTPUT.to_owned());
+        let lag_time = Some(register_histogram!(LAG_TIME_NAME));
```
Comment (tobz) [inline src/source_sender/mod.rs:93]: Same caveat here.

File: src/source_sender/mod.rs [hunk 3]
```diff
@@ -276,4 +321,125 @@ impl Inner {
...
+                    .timestamp()
+                    .map(|timestamp| timestamp.timestamp_millis()),
+                EventRef::Trace(trace) => trace
+                    .get(log_schema().timestamp_key())
+                    .and_then(get_timestamp_millis),
+            };
+            if let Some(timestamp) = timestamp {
+                // This will truncate precision for values larger than 2**52, but at that point the user
+                // probably has much larger problems than precision.
+                let lag_time = (reference - timestamp) as f64 / 1000.0;
```
Comment (tobz) [inline src/source_sender/mod.rs:344]: One thing that we may or may not want to do anything about is clock synchronization. If the system clock Vector uses is behind the clock used to generate the event's timestamp, we could end up emitting negative values here.... from the perspective of recording the metrics, we can _handle_ negative values, but they'll look weird downstream.

I'm not actually sure if it makes sense to only allow positive values, or to emit errors when we detect clock desync, or to increment a counter for clock desync errors, etc...

**Review by tobz — APPROVED** [review]
Generally, I'm +1 on these changes.

I left some comments/thoughts that aren't necessarily blocking, but might represent aspects that haven't been tested/fully fleshed out yet. I leave it up to you to decide if you feel like you need to explore any more of the functionality related to my comments.

File: src/source_sender/mod.rs [hunk 4]
```diff
@@ -276,4 +321,125 @@ impl Inner {
 
         Ok(())
     }
+
+    /// Calculate the difference between the reference time and the
+    /// timestamp stored in the given event reference, and emit the
+    /// different, as expressed in milliseconds, as a histogram.
+    fn emit_lag_time(&self, event: EventRef<'_>, reference: i64) {
+        if let Some(lag_time_metric) = &self.lag_time {
+            let timestamp = match event {
```
Comment (spencergilbert) [inline src/source_sender/mod.rs:330]: There are some formats that don't require an event timestamp, OTLP for example. We probably need to set a standard that we'd insert `now()` for missing timestamps.

File: src/source_sender/mod.rs [hunk 5]
```diff
@@ -200,19 +215,45 @@ impl SourceSender {
...
+            },
+            rx,
+        )
     }
 
     async fn send(&mut self, events: EventArray) -> Result<(), ClosedError> {
+        let reference = Utc::now().timestamp_millis();
+        events
+            .iter_events()
+            .for_each(|event| self.emit_lag_time(event, reference));
```
Comment (lukesteensen) [inline src/source_sender/mod.rs:256]: One thing we could consider is adding a timestamp field to the `EventArray` itself. There would be consistency concerns to think about, especially if event timestamps are changed within a codec or similar, but also a potentially meaningful performance improvement.

**Review by lukesteensen — APPROVED** [review]
My only real concern here is the performance drop, which would obviously be nice to avoid. 

I'm curious how much of the increased cost is the actual metric emission vs the work to calculate the values over each event, which would help indicate the best path towards reducing that cost. I like your idea of aggregating a value over the batch, but I might lean more towards a maximum than an average (though it is tempting to then add min, avg, etc until you start to approach a histogram again).

File: src/source_sender/mod.rs (hunk 1 above)
Comment (bruceg) [reply src/source_sender/mod.rs:88]: Part of this PR _is_ to run the builder inside the source component span.

File: src/source_sender/mod.rs (hunk 3 above)
Comment (bruceg) [reply src/source_sender/mod.rs:344]: Hmm, yes, that it something to highlight. This metric does assume that clocks are accurately synchronized.

File: src/source_sender/mod.rs (hunk 4 above)
Comment (bruceg) [reply src/source_sender/mod.rs:330]: Wouldn't that be an unrelated change, possibly related to schemas, as it changes the data model for those sources?

**Comment by bruceg** [comment]
> My only real concern here is the performance drop, which would obviously be nice to avoid.

With this last iteration, the performance change is within the soak test noise threshold.

File: src/source_sender/mod.rs (hunk 5 above)
Comment (bruceg) [reply src/source_sender/mod.rs:256]: I'm not sure how that would work if the events composing the array have differing timestamps.

File: src/source_sender/mod.rs (hunk 4 above)
Comment (spencergilbert) [reply src/source_sender/mod.rs:330]: Yeah, probably? I can't think of many other places where we're relying on the timestamp to be there (aside from sinks), just wanted to make sure this won't blow up if the field doesn't exist.

File: src/source_sender/mod.rs (hunk 1 above)
Comment (tobz) [reply src/source_sender/mod.rs:88]: Right you are. I did not review closely enough. :)

File: website/cue/reference/components/sources/internal_metrics.cue [hunk 6]
```diff
@@ -826,6 +826,12 @@ components: sources: internal_metrics: {
 			default_namespace: "vector"
 			tags:              _component_tags
 		}
+		lag_time_seconds: {
```
Comment (jszwedko) [inline website/cue/reference/components/sources/internal_metrics.cue:829]: @bruceg should this be `source_lag_time_seconds`?

**Pull request merged**